### PR TITLE
Fix problems with Conda path substitutions in binaries.

### PR DESCRIPTION
**Pull request by heplesser**
This PR replaces #2341 with a direct solution to the problems cause by binary path substitution during conda installation. It is based on https://github.com/conda/conda-build/issues/1674#issuecomment-280378336.

I have not tested it with the Conda build process yet.

### CODE REVIEW

File: nestkernel/dynamicloader.cpp [hunk 1]
```diff
@@ -270,7 +270,13 @@ DynamicLoaderModule::init( SLIInterpreter* i )
     LOG( M_ERROR, "DynamicLoaderModule::init", "Could not initialize libltdl. No dynamic modules will be available." );
   }
 
-  if ( lt_dladdsearchdir( NEST_INSTALL_PREFIX "/" NEST_INSTALL_LIBDIR ) )
+  // To avoid problems due to string substitution in NEST binaries during
+  // Conda installation, we need to convert the literal to string to be able to see it as
+  // a c_str; converting that to string again will limit at the first 0-byte in the cstr.
+  // We then need to convert back to c_str since lt_dladdsearchdir needs that.
+  // see #2237 and https://github.com/conda/conda-build/issues/1674#issuecomment-280378336
```
Comment (jougs) [inline nestkernel/dynamicloader.cpp:277]: I would shorten the comment to be the same as the one in `slistartup.cc`.

File: nestkernel/dynamicloader.cpp [hunk 2]
```diff
@@ -270,7 +270,13 @@ DynamicLoaderModule::init( SLIInterpreter* i )
...
+  const std::string module_dir = std::string( std::string( NEST_INSTALL_PREFIX ).c_str() ) + "/" + NEST_INSTALL_LIBDIR;
```
Comment (jougs) [inline nestkernel/dynamicloader.cpp:278]:
```suggestion
  const std::string module_dir = std::string( NEST_INSTALL_PREFIX ).c_str() + std::string( "/" NEST_INSTALL_LIBDIR );
```

**Review by jougs — COMMENTED** [review]
Only cosmetic suggestions from my side.

**Comment by jougs** [comment]
@steffengraber: can you please test the conda build (without my changes)? Thanks!

**Comment by heplesser** [comment]
@steffengraber I assume with NEST 3.2, you only get the 
```
Mar 18 11:04:32 SLIStartup [Fatal]: 
    SLI initialisation file not found at 
    $PREFIX
```
message, not the failed assertion from `scanner.cc:582`?

Is there any way I could look at the full build logs, the `libnestutil.h/config.h` file from the build directory and the binaries created, esp ´lib/libsli.so´?

File: nestkernel/dynamicloader.cpp (hunk 1 above)
Comment (heplesser) [reply nestkernel/dynamicloader.cpp:277]: Done.

**Comment by steffengraber** [comment]
That works so far. Great.  I have prepared the files for you again.

**Comment by heplesser** [comment]
@steffengraber @jougs I have now removed the debugging output. Since all seems to work well (the latest package works well on my Ubuntu 20 machine), I hope you can review quickly and we can get this merged.

I do not see any performance problems from compiling `slistartup.cc` and `dynamicloader.cpp` with `-O0` since code there is very rarely executed.

**Review by jougs — APPROVED** [review]
Wouldn't compiling with `-O0` have solved the problem already without any changes to actual code? I must say I am not totally happy with this solution, but willing to accept it, given the amount of time we already sunk in this...

And a short heads-up: The code in the `DynamicLoader` might change again when I get to addressing the remaining parts of #1717.

**Review by steffengraber — APPROVED** [review]
:+1:  All tests run smoothly again.

**Comment by isuruf** [comment]
This is an ugly hack whereas https://github.com/nest/nest-simulator/pull/2338 is a better solution. What's the reason for not using #2338?

**Comment by jougs** [comment]
First, the solution in #2338 did not work, as it composed paths in the wrong order. The main problem, however comes when NEST is compiled statically (which we do often on large supercomputers)  and the library path cannot be determined in the way you propose. I'd rather have a solution that does not depend on either of these and will look again into it when addressing #1717.

**Comment by isuruf** [comment]
> First, the solution in https://github.com/nest/nest-simulator/pull/2338 did not work, as it composed paths in the wrong order.

I think #2338 didn't work because it was incomplete (i.e. it didn't fix `dynamicloader.cpp` file)

> The main problem, however comes when NEST is compiled statically (which we do often on large supercomputers) and the library path cannot be determined in the way you propose.

Since this is a conda-build problem and we only build shared libraries, maybe #2338 can be done with an `#ifdef SUPPORT_CONDA_BUILD` flag or something similar.

**Comment by jougs** [comment]
The [concatenation of strings](https://github.com/nest/nest-simulator/pull/2338/files#diff-02ede1f3bf93d9d73a11a5d07fe3114214255afd2b0a15da14d9ab17e15eb7acR152) in #2338 was clearly the wrong way around. I printed the resulting value and it came out wrong.

We are currently in the process of phasing out SLI and the solution would hopefully not stay for long anyway. For the dynamic loader, things will anyway change quite a bit in the context of #1717 and I'm not sure it's worth to go for a cleaner solution right now, even though I agree that the one in this PR is ugly.

**Pull request merged**